File: csharp/Patterns/Singleton.cs
using System;
using System.Collections.Generic;

namespace DesignPatterns.Patterns;

public static class Singleton
{
    private sealed class Settings
    {
        private static readonly Settings InstanceValue = new();
        private readonly Dictionary<string, string> _values = new();

        private Settings() { }

        public static Settings Instance => InstanceValue;

        public void Set(string key, string value) => _values[key] = value;

        public string? Get(string key)
        {
            return _values.TryGetValue(key, out var value) ? value : null;
        }
    }

    public static void Run()
    {
        Console.WriteLine("\nSingleton");
        var settingsA = Settings.Instance;
        var settingsB = Settings.Instance;
        settingsA.Set("theme", "blue");
        Console.WriteLine($"Same instance: {ReferenceEquals(settingsA, settingsB)}");
        Console.WriteLine($"Theme: {settingsB.Get("theme")}");
    }
}

File: python/patterns/singleton.py
class Settings:
    _instance = None

    def __new__(cls) -> "Settings":
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.values = {}
        return cls._instance

    def set(self, key: str, value: str) -> None:
        self.values[key] = value

    def get(self, key: str) -> str | None:
        return self.values.get(key)


def run() -> None:
    print("\nSingleton")
    settings_a = Settings()
    settings_b = Settings()
    settings_a.set("theme", "blue")
    print("Same instance:", settings_a is settings_b)
    print("Theme:", settings_b.get("theme"))
